asf_installer_survey/notebooks/section_6_analysis.py

```diff
@@ -0,0 +1,1670 @@
+# - SQ: 42a, 42b
+# - Sample:
+# - Company_size_owner
+# - Company size_employee
+
+# %%
+# Generate crosstab dataframe
+df_602_crosstab_owner_V2 = crosstable("CompanySizeOwnerV2",
+                                      data,
+                                      "SQ42",
+                                      ans_42
+                                      )
+
+# Save to .csv and display
+df_602_crosstab_owner_V2.to_csv("../../outputs/section6/csv/602_owner_V2.csv")
+df_602_crosstab_owner_V2
+
+# %%
+# Generate figure
+groupedbar(df_602_crosstab_owner_V2,
+           "602_owner_V2",
+           "SQ 42a. Do you offer repairs, servicing and maintenance to heat pump customers?"
+           )
+
+# %%
+# Generate figure
+donut(df_602_crosstab_owner_V2,
+           "602_owner_V2",
+           "SQ 42a. Do you offer repairs, servicing and maintenance to heat pump customers?"
+           )
+
+# %%
+# Generate crosstab dataframe
+df_602_crosstab_employee_V2 = crosstable("CompanySizeEmployeeV2",
+                                        data,
+                                        "SQ42",
+                                        ans_42
+                                        )
+# Save to .csv and display
+df_602_crosstab_employee_V2.to_csv("../../outputs/section6/csv/602_employee_V2.csv")
+df_602_crosstab_employee_V2
+
+# %%
+# Generate figure
+groupedbar(df_602_crosstab_employee_V2,
+           "602_employee_V2",
+           "SQ 42b. Does the business you work for offer repairs, servicing and maintenance to heat pump customers?"
+           )
+
+# %%
+# Generate figure
+donut(df_602_crosstab_employee_V2,
+           "602_employee_V2",
+           "SQ 42b. Does the business you work for offer repairs, servicing and maintenance to heat pump customers?"
+           )
+
+# %% [markdown]
+# ### RQ 6.03 Do survey participants offer heat pump repairs, servicing and maintenance for all heat pump systems, or just ones they’ve installed?
+# - SQ: 43a, 43b, 43c
+# - Sample: Standard_output
+
+# %%
+# Print possible answers
+data[col.q43a].value_counts()
+data[col.q43b].value_counts()
+data[col.q43c].value_counts()
+
+# %%
+def condition_q43(x):
+    """
+    Function to be applied to dataframe which creates a new column
+    merging responses from sub-types of the same question.
+
+    Args:
+        x: pandas dataframe containing responses from analytical sample.
+    """
+    
+    # Merge responses    
+    if "Only" in x[col.q43a] or "Only" in x[col.q43b] or "Only" in x[col.q43c]:
+        return "Only for heat pump systems that I/my company/the business has installed"
+    
+    elif "For all" in x[col.q43a] or "For all" in x[col.q43b] or "For all" in x[col.q43c]:
+        return "For all heat pump systems, regardless of who originally installed it"
+    
+    else:
+        return None
+    
+# Apply function to each row of dataframe
+data["SQ43"] = data.apply(condition_q43,
+                          axis=1
+                          )
+
```

Comment:
Suggestion around a more concise way of writing out this function potentially:
```
from typing import Dict, Any, Optional, List

def condition_q43(x: Dict[str, Any], columns: List[str]) -> Optional[str]:
    """
    Function to be applied to dataframe which creates a new column
    merging responses from sub-types of the same question.

    Args:
        x: A dictionary-like object representing a single row from a DataFrame, 
           where the keys are column names and the values are the corresponding cell values.
        columns: A list of column names to check.
    """
    
    # Merge responses    
    for col in columns:
        if "Only" in x[col]:
            return "Only for heat pump systems that I/my company/the business has installed"
        elif "For all" in x[col]:
            return "For all heat pump systems, regardless of who originally installed it"
    
    return None

data['new_column'] = data.apply(condition_q43, columns=[col.q43a, col.q43b, col.q43c], axis=1)
```

Reply:
Ah thank you, it didn't occur to me initially to have the column numbers as an input the function!